### ISSUE TITLE
Scylla shard-aware Java driver throughput is limited compared to non-shard-aware driver

### ISSUE THREAD

**Issue by roydahan** [scylladb/scylladb#8996]
Comparing throughput of scylla using the shard aware driver (java) and non-shard-aware driver, resulted in about 25% less throughput (write and read) when using the shared aware version.

Using c-s with scylla-driver: scylla-driver-core-3.10.1-scylla-0-shaded.jar, max write throughput is about **75K ops**.
Using c-s with cassandra-driver: cassandra-driver-core-3.10.2.jar, max write throughput is about **100K ops**.

The test uses 4 loaders, that each run 2 processes (total of 8 c-s processes).
While using the shard-aware driver, it looks like above 4 processes there is no change in throughput at all (4 processes already produces 75K ops).
Using the stock Cassandra driver, when adding the 4 more processes, the throughput is going up to 100K ops.

*Installation details*
Scylla version (or git commit hash): 4.6.dev-0.20210531.b0c22f2e39
Cluster size: 3
OS (RHEL/CentOS/Ubuntu/AWS AMI): AWS AMI (ami-033f383abe0304292 us-east-1)

<img width="671" alt="Screen Shot 2021-07-08 at 20 21 08" src="https://user-images.githubusercontent.com/20959584/124964757-0c078600-e02a-11eb-99d7-d2a1abdccf15.png">

Live screenshot: https://snapshot.raintank.io/dashboard/snapshot/T3Sehx1yLtPIhlkgsOUfUoJMPz1EtCqc

Graphs explained:
1. First run is shard-aware loader (annotation 1).
2. Annoatation mumber 2 points on the time that adding 4 more process to that load didn't increase the throughput from 75K to 100K ops.
3. Second run, this time with cassandra-driver - first 4 processes producing 75K ops (annotation 3).
4. Annotation 4  points on the time of adding 4 more processes, showing how throughput reached to 100K ops.

Used the following staing job for my testing: https://jenkins.scylladb.com/job/scylla-staging/job/roy/job/NEW-scylla-master-perf-regression-aws-test-throughput/

Easy way to reproduce manually:
- Switch between drivers by going to /usr/share/scylla/cassandra/lib and switch between scylla and cassandra driver.

**Comment by avelanarius** [comment]
My first attempt at reproducing this (or quite similar) scenario:

1. Start 3 i3.2xlarge nodes with Scylla AMI (4.4.3; ami-09b29bd4c673ace35)
2. Start 2 c5.4xlarge loaders with Ubuntu 20.04 (ami-0996d3051b72b5b2c).
3. On each of the loaders run: `cassandra-stress write cl=QUORUM n=34285714 -pop seq=34285715..68571428 -schema "replication(strategy=SimpleStrategy,replication_factor=3)" -mode native cql3 -rate "threads=300 throttle=50000/s" -node 172.31.42.231,172.31.38.177,172.31.44.66` (modifying the `-pop seq` parameter per each loader)

As I was using Scylla Stress Orchestrator to start it up, these are the ways `cassandra-stress` was installed:
- Shard-aware: 
  ```
  sudo apt-get install -y apt-transport-https
  sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 5e08fbd8b5d6ec9c
  sudo curl -L --output /etc/apt/sources.list.d/scylla.list http://downloads.scylladb.com/deb/ubuntu/scylla-4.3-$(lsb_release -s -c).list
  sudo apt-get update -y
  sudo apt-get install -y scylla-tools
  ```
- Non-shard-aware:
  ```
  wget -q -N https://mirrors.netix.net/apache/cassandra/3.11.10/apache-cassandra-3.11.10-bin.tar.gz
  tar -xzf apache-cassandra-3.11.10-bin.tar.gz
  ```

Shard-aware, one of the loaders statistics (the total throughput is double the throughput of one loader):
```
Op rate                   :   32,217 op/s  [WRITE: 32,217 op/s]
Partition rate            :   32,217 pk/s  [WRITE: 32,217 pk/s]
Row rate                  :   32,217 row/s [WRITE: 32,217 row/s]
Latency mean              :    9.3 ms [WRITE: 9.3 ms]
Latency median            :    2.7 ms [WRITE: 2.7 ms]
Latency 95th percentile   :    8.5 ms [WRITE: 8.5 ms]
Latency 99th percentile   :  177.5 ms [WRITE: 177.5 ms]
Latency 99.9th percentile :  214.0 ms [WRITE: 214.0 ms]
Latency max               :  271.6 ms [WRITE: 271.6 ms]
Total partitions          : 34,285,714 [WRITE: 34,285,714]
Total errors              :          0 [WRITE: 0]
Total GC count            : 0
Total GC memory           : 0.000 KiB
Total GC time             :    0.0 seconds
Avg GC time               :    NaN ms
StdDev GC time            :    0.0 ms
Total operation time      : 00:17:44
```

Non-shard-aware, one of the loaders statistics:
```
Results:
Op rate                   :   43,799 op/s  [WRITE: 43,799 op/s]
Partition rate            :   43,799 pk/s  [WRITE: 43,799 pk/s]
Row rate                  :   43,799 row/s [WRITE: 43,799 row/s]
Latency mean              :    6.6 ms [WRITE: 6.6 ms]
Latency median            :    6.0 ms [WRITE: 6.0 ms]
Latency 95th percentile   :   12.0 ms [WRITE: 12.0 ms]
Latency 99th percentile   :   16.6 ms [WRITE: 16.6 ms]
Latency 99.9th percentile :   23.0 ms [WRITE: 23.0 ms]
Latency max               :  155.8 ms [WRITE: 155.8 ms]
Total partitions          : 34,285,715 [WRITE: 34,285,715]
Total errors              :          0 [WRITE: 0]
Total GC count            : 0
Total GC memory           : 0.000 KiB
Total GC time             :    0.0 seconds
Avg GC time               :    NaN ms
StdDev GC time            :    0.0 ms
Total operation time      : 00:13:02
```

That's a 26% drop in throughput.

I did not test different number of loaders, different threads count. The default `scylla.conf` on those Scylla AMI instances was used (so no `compaction_static_shares` changing; compactions could have (and probably did) change the results).

cc: @haaawk

**Comment by haaawk** [comment]
@avelanarius has performed more benchmarks and his observation is that most of the time shard-aware driver is better. The only case when non shard-aware driver is better is when you heavily overload the cluster. And I meant heavily not just a little bit. This is due to the fact that with cluster pushed way past its limits it turns out to be more effective to send a request to the least utilised shard instead of to the "right" shard. 

My conclusion would be that this is ok and the cluster shouldn't be pushed so much above its limits. During normal operation shard-aware driver proves to be better. That means we probably can close this issue. @slivne @avikivity what's your take on this?

**Comment by haaawk** [comment]
Also many thanks to @avelanarius for jumping on this issue and quickly getting to the bottom of it.

**Comment by avikivity** [comment]
The non-shard-aware driver doesn't know how to push the request to a less-loaded shard, so I don't understand the explanation.

**Comment by avikivity** [comment]
With fewer connections, the non-shard-aware has more batching compared to the shard-aware driver, but is also susceptible to overloading a shard and needs to do more work to move requests to the right shard.

**Comment by haaawk** [comment]
> The non-shard-aware driver doesn't know how to push the request to a less-loaded shard, so I don't understand the explanation.

It does. Driver keeps track of the number of inflight request on each connection and sends new request to the least loaded connection.

**Comment by avelanarius** [comment]
> With fewer connections, the non-shard-aware has more batching compared to the shard-aware driver, but is also susceptible to overloading a shard and needs to do more work to move requests to the right shard.

In my testing, I increased the number of connections on non-shard-aware driver (`connectionPerHost`) to be equal to shard-aware driver (confirmed by Grafana metrics). In the original Roy's run, the number of connections between drivers was also equal (8 shard Scylla and default 8 connections per host for non-shard-aware driver).

**Comment by avelanarius** [comment]
Here's a full write-up of my measurements:

### "Small" loaders
Scylla: 1 node i3.4xlarge, Scylla 4.4.3 AMI
Loaders: 2x c5.4xlarge, Ubuntu 20.04 LTS

Non-shard-aware driver version: 3.7.1
Shard-aware driver version: 3.7.1-scylla-2
Cassandra-stress version: scylla-tools-java as of 501332182308bfc7ad7abbd3efcf5b7ae19d3906

I ran two identical clusters concurrently and two identical sets of loaders (4 loaders total). The first cluster was loaded with non-shard-aware driver, while the other was loaded with shard-aware driver - exactly the same loads, started at the same time. I was gradually increasing the `throttle` parameter of `cassandra-stress`:

`cassandra-stress  write duration=15m -mode native cql3 connectionsPerHost=14 -rate "threads=500 throttle=10000/s" -node 127.0.0.2` 

The number of connections is the same between shard-aware and non-shard-aware driver (tuned with `connectionsPerHost=14`).

#### 120k/s

##### Non-shard-aware

![120k/s non-shard-aware](https://user-images.githubusercontent.com/10003183/126148818-7cfb0fbe-7b91-43d7-b72e-4ba305bc75e5.png)

##### Shard-aware

![120k/s shard-aware](https://user-images.githubusercontent.com/10003183/126148947-ee437654-0fce-4ad6-a507-56ff8a402042.png)

#### 200k/s

##### Non-shard-aware

![200k/s non-shard-aware](https://user-images.githubusercontent.com/10003183/126149123-23f6afd1-c0cd-4f62-886a-2e303ef61213.png)

##### Shard-aware

![200k/s shard-aware](https://user-images.githubusercontent.com/10003183/126149380-167cdd92-3940-48f3-9fd4-117513fc8615.png)

#### Conclusions

As you can see from those runs, shard-aware driver has smaller latencies (especially 99%) and Scylla load is smaller (~8%) with shard-aware driver. 

I did however observe a larger CPU usage on loaders with shard-aware driver:
 - 5 minute load average on non-shard-aware loader: 21.08
 - 5 minute load average on shard-aware loader: 25.93

(cassandra-stress was running for more than 5 minutes and for the same time on those two loaders) 

This higher CPU usage caused the shard-aware loader instance (not Scylla) to be overloaded with higher throughputs, while the non-shard-aware driver was not, which limited the maximum throughput I could achieve. Therefore I continued the tests with larger loaders. I did not investigate the reason for the higher loader CPU usage.

Raw data: [runs_small_loaders.zip](https://github.com/scylladb/scylla/files/6840413/runs_small_loaders.zip)

### "Large" loaders
The same setup, except for loaders:
2x c5n.9xlarge, Ubuntu 20.04 LTS

#### 240k/s

##### Non-shard-aware
![240k/s non-shard-aware](https://user-images.githubusercontent.com/10003183/126150957-5c9f2a81-5828-457b-abe8-4427dcfc7938.png)

##### Shard-aware
![240k/s shard-aware](https://user-images.githubusercontent.com/10003183/126151098-b12b7247-d293-4ce0-b102-e2570fc1b557.png)

#### 340k/s

##### Non-shard-aware
![340k/s non-shard-aware](https://user-images.githubusercontent.com/10003183/126151300-213b7025-9f15-4c25-a8d8-2159004d5798.png)

##### Shard-aware
![340k/s shard-aware](https://user-images.githubusercontent.com/10003183/126151438-21e2bf10-466b-4912-bb72-6bdb812defc3.png)

#### 400k/s

##### Non-shard-aware
![400k/s non-shard-aware](https://user-images.githubusercontent.com/10003183/126152259-e60c5ba0-0977-4072-aebc-cbd680d1646b.png)

##### Shard-aware
![400k/s shard-aware](https://user-images.githubusercontent.com/10003183/126153345-f9468418-6dd5-4ac3-8864-99edefa36560.png)

#### 458k/s, 410k/s (Scylla unable to hit a higher throttle limit)

##### Non-shard-aware
![458k/s non-shard-aware](https://user-images.githubusercontent.com/10003183/126152650-c6c611fa-c947-415d-87d0-51f023ba2d46.png)

##### Shard-aware
![410k/s shard-aware](https://user-images.githubusercontent.com/10003183/126152954-2175949e-0036-4d60-be46-5b71b516d195.png)

#### Conclusions
Until 340k/s, the shard-aware driver was still winning. However, with a higher load than that, the non-shard-aware driver was able to achieve better latencies. Finally, the non-shard-aware driver was able to achieve higher maximum throughput.

Keep in mind that this maximum throughput is nearly double (1.9x) the throughput when Scylla was under a 97% load (240k/s).

Raw data: [runs_large_loaders.zip](https://github.com/scylladb/scylla/files/6840419/runs_large_loaders.zip)

**Comment by psarna** [comment]
At your leisure, it would be nice to rerun the experiment and also collect each node's logs from the cases where the shard-aware driver started having degraded performance compared to non-shard-aware. Perhaps it will allow us to pinpoint the problem and fix it.

**Comment by avikivity** [comment]
What's worrying me is that the non-shard-aware driver achieves higher throughput. Latency at 100% utilization is not an important metric like latency at lower utilizations, but throughput is.

**Comment by roydahan** [comment]
This is a throughput test, so obviously one who after throughput is trying to maximize the throughput he gets from the cluster.
In that case, one will try to push cluster to its limits and wouldn't care about CPU utlization or what's going on with the shards.
The latency, as @avikivity mentioned, in these cases aren't that relevant.

**Comment by slivne** [comment]
I don't understand why the throughput of the shard aware driver is lower the non shard aware driver.

Putting aside the additional stress on the client to compute the correct shard.

If the distribution of all the data over all the shards is the same then and the distribution of requests is the same - why would the shard aware receive less traffic.

We need to look at shard view and not in node view - we may have a slow "shard".

Looking at shard view we will be able to see if all the shards get ~the same number of requests (data distribution between shards) and if we have a slow shard slowing down the client with queued requests.

Is there any difference on the queuing / amount of parallel requests we will send on a single connection ?

One thing that maybe be different is if buffering of requests is applied (nagel's algorithm)on the client side - on the shard aware driver items are queued into connections based on their shard and maybe they wait longer ?

**Comment by haaawk** [comment]
I talked about it with @vladzcloudius and even implemented o PoC in gocql that switches from shard awarness to the old behaviour when the node becomes busy. https://github.com/scylladb/gocql/pull/86

**Comment by roydahan** [comment]
With recent scylla versions, we see that shard-aware driver has better throughput than stock java driver.
I'll update with details soon.

**Comment by avelanarius** [comment]
> > With recent scylla versions, we see that shard-aware driver has better throughput than stock java driver. I'll update with details soon.
> 
> Is there any change in Scylla Java driver that could explain this, @roydahan ?

I believe he meant some change in Scylla (database, not driver). There isn't any change in the driver I'm aware of that would improve the performance. And also if the improvement was seen on `cassandra-stress` from `scylla-tools-java`, until just recently it used a very old Scylla Java Driver (from 2019).

**Comment by vladzcloudius** [comment]
> > > With recent scylla versions, we see that shard-aware driver has better throughput than stock java driver. I'll update with details soon.
> > 
> > 
> > Is there any change in Scylla Java driver that could explain this, @roydahan ?
> 
> I believe he meant some change in Scylla (database, not driver). There isn't any change in the driver I'm aware of that would improve the performance. And also if the improvement was seen on `cassandra-stress` from `scylla-tools-java`, until just recently it used a very old Scylla Java Driver (from 2019).

Got it. 
@avelanarius is there any progress with the Java version of the patch mentioned by PiotrJ?

**Comment by roydahan** [comment]
@avelanarius apparently the shard-aware is still a bit worse in terms of throughput.
If you have the priority and capacity to work on this issue, we can share some tests to reproduce and measure.

**Comment by Gor027** [comment]
An attempt to reproduce a similar scenario:

* Setup a Scylla cluster with 5 `i3.4xlarge` nodes with Scylla 5.1.3 (AMI-07d0ff1841b1a9d8a)
* Start 2 `c5n.9xlarge` loaders with Ubuntu 22.04 LTS (AMI-0ab0629dba5ae551d)
* Run on each loader a c-s process with the following command:
    `cassandra-stress write duration=150m -schema 'replication(strategy=SimpleStrategy, factor=3)' -mode native cql3 -rate "threads=500 connectionsPerHost=14" -pop 'dist=UNIFORM(1..1000000000000)'  -node NODE_IP`

###### All the nodes and loaders were in `us-east-2a`

<details open>
  <summary>Differences with the previous benchmarks</summary>

- Cluster contains 5 nodes
- Replication factor 3
- Consistency level: LOCAL ONE
- The loaders sent only write queries and they were not throttled
- Queries were distributed in a larger range (1 ... 10^12) so the dataset did not fit in the RAM

</details>

### Shard-aware driver (Version: [GitHub master](https://github.com/scylladb/java-driver))

![image](https://user-images.githubusercontent.com/29541844/216387213-d49c61e5-1b0e-48a3-a734-2df654b9fe3b.png)

### Non-shard-aware driver (Version: [Apache Cassandra 3.11.14](https://dlcdn.apache.org/cassandra/3.11.14/apache-cassandra-3.11.14-bin.tar.gz))

![image](https://user-images.githubusercontent.com/29541844/216390275-54f4f147-d4f0-41be-817c-bd3b9d70eae8.png)

The benchmark results have shown that overall the throughput is the same for shard-aware and non-shard-aware drivers. However, using the shard-aware driver resulted in timeouts which did not happen while using the non-shard-aware driver:

![image](https://user-images.githubusercontent.com/29541844/216392230-04b1badd-f2c5-4657-a487-3da9875f25ec.png)

Perhaps, this kind of benchmark is not so "realistic", as the dataset does not fit in the RAM, and the cluster gets easily overloaded with background processes like compaction, hints which result in high latencies and eventually timeouts.

### Conclusion

The benchmarks showed that the shard-aware driver is overall better than the non-shard-aware driver when the cluster is not under a high load. Even in this benchmark, you can notice that at the beginning the shard-aware driver had a higher spike in throughput than the non-shard-aware driver but gradually the throughput decreased to a certain limit for both drivers because other background jobs used a big percentage of CPU time.
To sum up, the benchmarks showed that in most cases, the shard-aware driver performs better and provides higher throughput than the non-shard-aware driver. and with an overloaded cluster, in terms of throughput, it is not inferior to it. So, maybe this issue can be closed now, what do you think? @avelanarius @roydahan @vladzcloudius

**Comment by avelanarius** [comment]
Note that those results are separate matter to https://github.com/scylladb/java-driver/issues/136. This issue deals with “max throughput”, while the linked issue is more about latency spikes. We will soon write a response on that issue too, including the test @vladzcloudius suggested: https://github.com/scylladb/java-driver/issues/136#issuecomment-1324384204

It could be true that this issue can be closed, while the linked one stills needs a fix (which we have).

**Comment by roydahan** [comment]
We will do another run with latest Scylla and our test and let you know if we see the same things.

**Comment by dorlaor** [comment]
Agree that the 100% utilization matters less and it's better to measure at
75% load.

Also interesting is to measure normal(Zipfian) distribution, since shard
aware
should handle it better

Lastly, why use the old Java branch - 3.x, why not the new 4.y one?

**Comment by vladzcloudius** [comment]
On 2/2/23 8:00 PM, Dor Laor wrote:
> Agree that the 100% utilization matters less and it's better to measure at
> 75% load.

It depends on what you want to test exactly.
If you want to find the maximum throughput (which I assume *is* the case 
here based on the subject) you should push as much as the system can 
deliver until you see timeouts.
We are looking at the maximum throughput the system can deliver without 
timeouts.

And the CPU utilization in this case is just one of many metrics - you 
can start getting timeouts at 40% CPU utilization in one case and you 
can be able to add more throughput after you reached 100% in another.

>
> Also interesting is to measure normal(Zipfian) distribution, since shard
> aware
> should handle it better
>
> Lastly, why use the old Java branch - 3.x, why not the new 4.y one?
>

**Comment by aleksbykov** [comment]
Performance tests were run for latest master:
Version Details:
build date: 20230207
commit id: a5ff375a74e7
build-id: 7cbb9e687d4354d14cfbcb86c7f2aa30bb88e3d6

## For mixed test we have very similar results:
With non-shard aware driver:
![Screenshot from 2023-02-14 14-21-57](https://user-images.githubusercontent.com/44991177/218666853-a72799ee-6634-417a-b2e0-591118e9303a.png)

_Restore monitor: https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?sct_branch=master&test_id=e2dc214c-34ce-424d-9470-4a3e05108068_


With shard aware:
![Screenshot from 2023-02-14 14-22-44](https://user-images.githubusercontent.com/44991177/218667074-815c6cea-d1d5-45f5-9b2b-cd4650039872.png)

_Restore monitor: https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?sct_branch=master&test_id=e20b360c-2107-4670-8af1-63b96ad36410_


## for read test 
with shard aware driver
![Screenshot from 2023-02-14 14-32-12](https://user-images.githubusercontent.com/44991177/218669112-cd735bc7-2b3c-4353-a140-9182d9c25700.png)
_Restore monitor: https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?sct_branch=master&test_id=8b8606c5-bf31-4210-9f6b-d5ec2a440d8f_


with non-shard aware driver
![Screenshot from 2023-02-14 14-30-36](https://user-images.githubusercontent.com/44991177/218668822-cdbcafaf-364b-4e76-901f-d8a06f610e6b.png)

_Restore monitor: https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?sct_branch=master&test_id=a84e7af6-145e-4bf4-b2a9-ebe076897bc2_


## for Write test

with shard aware
![Screenshot from 2023-02-14 14-33-58](https://user-images.githubusercontent.com/44991177/218669450-f05d9412-1469-42d9-89f0-9fc37377118a.png)

_Restore monitor: https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?sct_branch=master&test_id=eb6b804a-6af6-417b-bd02-afd2e0b50904_


with non-shard aware
![Screenshot from 2023-02-14 14-35-45](https://user-images.githubusercontent.com/44991177/218669888-cbe6dcf0-eb6e-4860-b38d-86b431bbcaf5.png)

_Restore monitor: https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?sct_branch=master&test_id=24b548af-c457-4a4b-a5f3-068a6f7f138f_

**Comment by mykaul** [comment]
(nit: it'd be easier to read in a table than from screenshots)

**Comment by roydahan** [comment]
@aleksbykov 
1. It's not readable like this, especially for people who don't know to read these reports.
2. These are results for enterprise branch, let's include also master branch.
3. We need to "qualify" the new jobs, as I see that the old test which is also "non shard-aware" produced 144K op/s while the new test got only "136K" op/s (test_write), read and mixed looks quite the same.

**Comment by aleksbykov** [comment]
@roydahan  here is a link to spreadsheet. https://docs.google.com/spreadsheets/d/1X3NToAwND2MhYW7reLaObdFJcKVsGmFiXQ3_EM_Qseo/edit?usp=sharing
Is format ok?

**Comment by roydahan** [comment]
Testing with recent drivers and latest scylla shows that now for shard aware the 100% write throughput is better than the non-shard aware, the 100% read is slightly worse and the mixed is the same.



job name | build date | write | read | mixed
-- | -- | -- | -- | --
throughput-shard-aware-test | build date: 20230207 commit id: a5ff375a74e7 | 173,742 | 154,824 | 107,499
throughput-non-shard-aware-test | build date: 20230207 commit id: a5ff375a74e7 | 149,016 | 160,737 | 108,342

**Comment by vladzcloudius** [comment]
@roydahan @mykaul Did you use a Java driver with the https://github.com/scylladb/java-driver/issues/136?

If not then these results make no sense.
This GH issue is about a known problem in the shard-aware driver. No change on scylla can resolve that.

cc @avelanarius

**Comment by vladzcloudius** [comment]
@avelanarius one of the main changes in the new Cassandra driver is that by default they open a single TCP socket to each node. Before it was 4.

@roydahan a few questions about the test above:
1) Are you sure you don't ignore errors? If there are timeouts during the test the test results are invalid.
2) How do you make sure the loader is not a bottleneck?
3) How do you make sure the results are reliable/consistent? 
  a) Do you enforce `sq_split`? On a latest `master` an `sq` would be auto-selected for 8 vCPU machine. When this issue was created it would auto-select `mq`. You were supposed to override it with `sq_split` both before and today.
  b) Do you pin cassandra-stress processes to specific CPUs on a loader? 


@Gor027 @aleksbykov 
General recommendations:
1) Never look at metrics other than in per-shard view. When you aggregate you miss the most interesting.
2) Always make sure that the coordinator processing is ideally (or very close to it) balanced.
3) Keep an eye on a "Foreground reads/writes" graphs. 
4) Always run a non-shard-aware clients with the same amount of TCP connections (on every client!) as shard-aware ones. Otherwise it's not apples-to-apples comparison.

Expected result: without https://github.com/scylladb/java-driver/issues/136 and Friends shard-aware driver would suck badly compared to non-shard-aware.

@mykaul I strongly recommend reopening this GH issue until we better understand a reason for a "mysterious improvement".

**Comment by vladzcloudius** [comment]
@DoronArazii @tomer-sandler FYI

**Comment by roydahan** [comment]
> @avelanarius one of the main changes in the new Cassandra driver is that by default they open a single TCP socket to each node. Before it was 4.
> 
> @roydahan a few questions about the test above:
> 
> 1. Are you sure you don't ignore errors? If there are timeouts during the test the test results are invalid.
Yes
> 2. How do you make sure the loader is not a bottleneck?
We used big enough loader and we have a test that checks what's the ideal number of loaders, num of c-s processes and num of threads per process.
But I don't see how it's related, or you're talking about the non-shard-aware loader?
Anyway, it wasn't overloaded.
> 3. How do you make sure the results are reliable/consistent?
>    a) Do you enforce `sq_split`? On a latest `master` an `sq` would be auto-selected for 8 vCPU machine. When this issue was created it would auto-select `mq`. You were supposed to override it with `sq_split` both before and today.
Not sure I understand, we use default scylla, here we compared the same scylla with and without shard aware driver.
#>    b) Do you pin cassandra-stress processes to specific CPUs on a loader?
Yes.
> 
> 
> @mykaul I strongly recommend reopening this GH issue until we better understand a reason for a "mysterious improvement".

For now, I don't see a reason to open it, since shard-aware give us even better throughput for this test.
The "mysterious improvement" can be anything that happened in the last 4 major scylla releases or in the last year of driver changes.
So, if you have a suggested test that can show that shard-aware is worse than non-shard-aware for throughput test, please suggest it.

**Comment by roydahan** [comment]
I don’t see which questions were not answered.

**Comment by vladzcloudius** [comment]
> I don’t see which questions were not answered.

Oh, I see you answered them. 
The formatting of your response leaves much to be desired however. Let me try respond to your responses then.

**Comment by vladzcloudius** [comment]
> @vladzcloudius the formatting is impossible to answer now with inlines. I'll summarize and if you want, open a new issue where you think there is an issue with shard aware.
> 
>     1. Loader isn't overloaded.
> 
>     2. We test here with i3.2xlarge.
> 
>     3. We don't change and won't change the default config scylla comes with sq, sq_split or whatever it will be next.

@roydahan Unless you address items I mentioned in my previous comments, which in particular refers to items (2) and (3) above your test results are unreliable.
